# Bearbeitung des Datensatzes "Inventory" der Seattle Public Library mit Jupyter Notebook

Aus dem Datensatz "Inventory" sollen in diesem Teil der Aufgabe die Bestandszahlen der Zentralbibliothek betrachtet werden:

- Wie hoch war der Bestand an Nonfiction-Büchern in der Zentralbibliothek der Seattle Public Library im Jahr 2018?

Eine Einschränkung hierbei ist, dass die Zeilenmenge für diese Aufgabe auf 3.000.000 begrenzt wird.
Darüber hinaus wird der Sammlungstyp auf "CA-Nonfiction" eingeschränkt.

## Frage 5: Wie hoch war der Bestand an Nonfiction-Büchern in der Zentralbibliothek der Seattle Public Library im Jahr 2018?

In [2]:
# Zunächst laden wir uns die benötigten Bibliotheken herunter
import urllib.request

In [3]:
import pandas as pd

In [29]:
import numpy as np

In [4]:
# Wir legen vor dem Download die Variablen für die url und die csv-Datei fest. Das Limit legen wir auch hier auf 3.000.000.
url="https://data.seattle.gov/resource/6vkj-f5xf.csv?$limit=3000000"
inv="inventory2018.csv"

In [5]:
# Wir können die Datei nun herunterladen
urllib.request.urlretrieve(url, inv)

('inventory2018.csv', <http.client.HTTPMessage at 0x22ae2a11e88>)

In [31]:
# Anschließend erstellen wir eine Variable zum Lesen der Datei und lassen sie uns anzeigen
invent=pd.read_csv(inv)
invent

bibnum                                              title  \
0        3011076  A tale of two friends / adapted by Ellie O'Rya...   
1        2248846  Naruto. Vol. 1, Uzumaki Naruto / story and art...   
2        3209270  Peace, love & Wi-Fi : a ZITS treasury / by Jer...   
3        1907265     The Paris pilgrims : a novel / Clancy Carlile.   
4        1644616  Erotic by nature : a celebration of life, of l...   
...          ...                                                ...   
2999995  2831460                 The Frank show / David Mackintosh.   
2999996  2983333  Eating dangerously : why the government can't ...   
2999997  3141286  How to make your money last : the indispensabl...   
2999998  2923232  Sock art : bold, graphic knits for your feet /...   
2999999   323272  Okanogan County agriculture, Washington [by] W...   

                                                author  \
0                                        O'Ryan, Ellie   
1                            Kishimoto, Masashi, 1974-   
2                                  Scott, Jerry, 1955-   
3                               Carlile, Clancy, 1930-   
4                                                  NaN   
...                                                ...   
2999995                              Mackintosh, David   
2999996                          Booth, Michael, 1965-   
2999997                             Quinn, Jane Bryant   
2999998                       Janssen, Edelgard, 1938-   
2999999  Washington (State). Department of Agriculture   

                                                      isbn publicationyear  \
0        1481425730, 1481425749, 9781481425735, 9781481...           2014.   
1                                               1569319006    2003, c1999.   
2                                144945867X, 9781449458676           2014.   
3                                               0786706155          c1999.   
4                                               094020813X    1991, c1988.   
...                                                    ...             ...   
2999995                          1419703935, 9781419703935           2012.   
2999996                          1442222662, 9781442222663          [2014]   
2999997  1476743762, 1476743770, 9781476743769, 9781476...           2016.   
2999998                          157076557X, 9781570765575          c2013.   
2999999                                                NaN          [1965]   

                                 publisher  \
0                         Simon Spotlight,   
1                                     Viz,   
2               Andrews McMeel Publishing,   
3                          Carroll & Graf,   
4        Red Alder Books/Down There Press,   
...                                    ...   
2999995    Abrams Books for Young Readers,   
2999996              Rowman & Littlefield,   
2999997                  Simon & Schuster,   
2999998                  Trafalgar Square,   
2999999                                NaN   

                                                  subjects itemtype  \
0        Musicians Fiction, Bullfighters Fiction, Best ...     jcbk   
1        Ninja Japan Comic books strips etc, Comic book...     acbk   
2        Duncan Jeremy Fictitious character Comic books...     acbk   
3        Hemingway Ernest 1899 1961 Fiction, Biographic...     acbk   
4        Erotic literature American, American literatur...     acbk   
...                                                    ...      ...   
2999995  Grandfathers Juvenile fiction, Show and tell p...     jcbk   
2999996  Food poisoning United States Prevention, Food ...     acbk   
2999997  Retirement income Planning, Retirees Finance P...     acbk   
2999998                           Knitting Patterns, Socks     acbk   
2999999       Agriculture Washington State Okanogan County     arbk   

        itemcollection floatingitem itemlocation               reportdate  \
0                ncrdr     Floating     

In [7]:
# Wir lassen uns die Titel der einzelnen Spalten anzeigen
invent.columns

Index(['bibnum', 'title', 'author', 'isbn', 'publicationyear', 'publisher',
       'subjects', 'itemtype', 'itemcollection', 'floatingitem',
       'itemlocation', 'reportdate', 'itemcount'],
      dtype='object')

In [32]:
# Für unsere Frage sind die Spalten "itemcollection" und "itemlocation" relevant. Wir kürzen die Darstellung auf diese Spalten.
invent[["itemcollection","itemlocation"]]

itemcollection itemlocation
0                ncrdr          qna
1              nycomic          lcy
2              nycomic          bea
3                cafic          cen
4                 canf          cen
...                ...          ...
2999995          ncpic          glk
2999996           canf          cen
2999997           nanf          cap
2999998           canf          cen
2999999           cs7r          cen

[3000000 rows x 2 columns]

In [33]:
# Wir interessieren uns für den Bestand der Zentralbibliothek ("cen"). 
# Deshalb filtern wir die Titel mit dieser Itemlocation heraus und lassen uns dies zunächst als Liste anzeigen.
invlist=invent.itemlocation.str.contains(pat="cen")
invlist

0          False
1          False
2          False
3           True
4           True
           ...  
2999995    False
2999996     True
2999997    False
2999998     True
2999999     True
Name: itemlocation, Length: 3000000, dtype: bool

In [34]:
# Die Liste zeigen wir nun als Tabelle an
invcen=invent[invlist]
invcen

bibnum                                              title  \
3        1907265     The Paris pilgrims : a novel / Clancy Carlile.   
4        1644616  Erotic by nature : a celebration of life, of l...   
5        1736505  Children of Cambodia's killing fields : memoir...   
6        1749492  Anti-Zionism : analytical reflections / editor...   
12       2519097  Stop bullying now! [videorecording] : Take a s...   
...          ...                                                ...   
2999989  3286348                Who's that girl / Blair Thornburgh.   
2999990  2918511                          All alone / Kevin Henkes.   
2999996  2983333  Eating dangerously : why the government can't ...   
2999998  2923232  Sock art : bold, graphic knits for your feet /...   
2999999   323272  Okanogan County agriculture, Washington [by] W...   

                                                author  \
3                               Carlile, Clancy, 1930-   
4                                                  NaN   
5                                                  NaN   
6                                                  NaN   
12                                                 NaN   
...                                                ...   
2999989                              Thornburgh, Blair   
2999990                                  Henkes, Kevin   
2999996                          Booth, Michael, 1965-   
2999998                       Janssen, Edelgard, 1938-   
2999999  Washington (State). Department of Agriculture   

                                                      isbn publicationyear  \
3                                               0786706155          c1999.   
4                                               094020813X    1991, c1988.   
5                                   0300068395, 0300078730          c1997.   
6                                               091559773X          c1989.   
12                                                     NaN         [2006?]   
...                                                    ...             ...   
2999989                          0062447777, 9780062447777          [2017]   
2999990  0060541156, 0060541164, 9780060541156, 9780060...          c2003.   
2999996                          1442222662, 9781442222663          [2014]   
2999998                          157076557X, 9781570765575          c2013.   
2999999                                                NaN          [1965]   

                                                 publisher  \
3                                          Carroll & Graf,   
4                        Red Alder Books/Down There Press,   
5                                   Yale University Press,   
6                                             Amana Books,   
12       U.S. Dept. of Health and Human Services ; Dept...   
...                                                    ...   
2999989  HarperTeen, an imprint of HarperCollinsPublish...   
2999990        Greenwillow Books/HarperCollins Publishers,   
2999996                              Rowman & Littlefield,   
2999998                                  Trafalgar Square,   
2999999                                                NaN   

                                                  subjects itemtype  \
3        Hemingway Ernest 1899 1961 Fiction, Biographic...     acbk   
4        Erotic literature American, American literatur...     acbk   
5        Political atrocities Cambodia, Children Cambod...     acbk   
6        Berger Elmer 1908 1996, Zionism Controversial ...     acbk   
12       Bullying in schools United States Prevention, ...    acdvd   
...                                                    ...      ...   
2999989  High school students Fiction, Rock musicians F...     acbk   
2999990                          Solitude Juvenile fiction     jcbk   
2999996  Food poisoning United States Prevention, Food ...     acbk   
2999998                           Knitting Patterns, Socks     acbk 

In [35]:
# Auch diese Tabelle reduzieren wir wieder auf die relevanten Spalten.
invcen[["itemcollection","itemlocation"]]

itemcollection itemlocation
3                cafic          cen
4                 canf          cen
5                 canf          cen
6                 canf          cen
12             cadvdnf          cen
...                ...          ...
2999989          cyfic          cen
2999990          ccpic          cen
2999996           canf          cen
2999998           canf          cen
2999999           cs7r          cen

[1206453 rows x 2 columns]

In [36]:
# Wir vergeben eine Variable für diese Tabelle, um sie bei weiteren Schritten als Grundlage nutzen zu können.
invallcen=invcen[["itemcollection","itemlocation"]]
invallcen

itemcollection itemlocation
3                cafic          cen
4                 canf          cen
5                 canf          cen
6                 canf          cen
12             cadvdnf          cen
...                ...          ...
2999989          cyfic          cen
2999990          ccpic          cen
2999996           canf          cen
2999998           canf          cen
2999999           cs7r          cen

[1206453 rows x 2 columns]

In [50]:
# In der Spalte "Itemcollection" wollen wir nun den Typ CA-Nonfiction ("canf") herausfiltern.
nonfall=invallcen.itemcollection.str.contains(pat="canf")
nonfall

3          False
4           True
5           True
6           True
12         False
           ...  
2999989    False
2999990    False
2999996     True
2999998     True
2999999    False
Name: itemcollection, Length: 1206453, dtype: bool

In [59]:
# Wir übertragen dieses Ergebnis in eine Tabelle und legen für diese eine Variable fest
invcanf=invcen[nonfall]
invcanf

bibnum                                              title  \
4        1644616  Erotic by nature : a celebration of life, of l...   
5        1736505  Children of Cambodia's killing fields : memoir...   
6        1749492  Anti-Zionism : analytical reflections / editor...   
27       2352640  Morningstar guide to mutual funds : five-star ...   
38         11879  Mirror to the American past; a survey of Ameri...   
...          ...                                                ...   
2999978    81740  Contexts of the drama [compiled by] Richard Go...   
2999980  1845598  Aviation, transport services : agreement betwe...   
2999983   196512  Selections from Carlyle: Sartor Resartus, The ...   
2999996  2983333  Eating dangerously : why the government can't ...   
2999998  2923232  Sock art : bold, graphic knits for your feet /...   

                                      author                       isbn  \
4                                        NaN                 094020813X   
5                                        NaN     0300068395, 0300078730   
6                                        NaN                 091559773X   
27                                       NaN  0471718327, 9780471718321   
38       Williams, Hermann Warner, 1908-1974                 0821204440   
...                                      ...                        ...   
2999978             Goldstone, Richard Henry                        NaN   
2999980                             Austria.                        NaN   
2999983           Carlyle, Thomas, 1795-1881                        NaN   
2999996                Booth, Michael, 1965-  1442222662, 9781442222663   
2999998             Janssen, Edelgard, 1938-  157076557X, 9781570765575   

        publicationyear                                          publisher  \
4          1991, c1988.                  Red Alder Books/Down There Press,   
5                c1997.                             Yale University Press,   
6                c1989.                                       Amana Books,   
27               c2005.                                John Wiley & Sons ,   
38               [1973]                           New York Graphic Society   
...                 ...                                                ...   
2999978          [1968]                                        McGraw-Hill   
2999980         [1998?]  Dept. of State : For sale by the Supt. of Docs...   
2999983         [c1915]                                   D.C. Heath & co.   
2999996          [2014]                              Rowman & Littlefield,   
2999998          c2013.                                  Trafalgar Square,   

                                                  subjects itemtype  \
4        Erotic literature American, American literatur...     acbk   
5        Political atrocities Cambodia, Children Cambod...     acbk   
6        Berger Elmer 1908 1996, Zionism Controversial ...     acbk   
27                                            Mutual funds     acbk   
38       Genre painting American, Genre painting 18th c...     acbk   
...                                                    ...      ...   
2999978                                              Drama     acbk   
2999980  Airlines Management International cooperation,...     acbk   
2999983                                                NaN     acbk   
2999996  Food poisoning United States Prevention, Food ...     acbk   
2999998                           Knitting Patterns, Socks     acbk   

        itemcollection floatingitem itemlocation               reportdate  \
4                 canf          NaN          cen  2017-09-01T00:00:00.000   
5                 canf          NaN          cen  2017-09-01T00:00:00.000   
6                 canf          NaN          cen  2017-09-01T00:00:00.000   
27                canf          NaN          cen  2017-09-01T00:00:00.000   
38                canf          NaN          cen  2017-09-01T00:00:00.000   
... 

In [58]:
# Nun reduzieren wir die Tabelle auf die für uns relevanten Spalten
canfcen=invcanf[{"itemcollection","itemlocation"}]
canfcen

itemlocation itemcollection
4                cen           canf
5                cen           canf
6                cen           canf
27               cen           canf
38               cen           canf
...              ...            ...
2999978          cen           canf
2999980          cen           canf
2999983          cen           canf
2999996          cen           canf
2999998          cen           canf

[428924 rows x 2 columns]

Lösung 5: Unsere Ergebnistabelle enthält 428924 Zeilen. Demnach befinden sich (unter Berücksichtigung der von uns festgelegten Einschränkungen) __428924__ Titel des Sammlungstyps CA-Nonfiction im Bestand der Zentralbibliothek der Seattle Public Library.